casper/src/main/resources/Registry.rho

```diff
@@ -330,104 +358,241 @@ in {
-                for (@response <- responseCh) {
-                  match response {
-                    Nil => {
-                      new ack in {
-                        TreeHashMap!("set", map, uri, (nonce, data), *ack) |
-                        for (_ <- ack) {
-                          ret!(uri)
+      } |
+      // Registers `insertSignedCh` as an insertSigned channel in the registry
+      bootstrapInsertSigned!(*insertSignedCh) | // this will work only once
+      for (insertSigned <- insertSignedCh) {
+        // Verifies signature and inserts data into registry
+        // Returns generated uri on `ret`
+        contract insertSigned(@(nonce, data), deployerID, ret) = {
+          new uriCh, hashCh, verifyCh, deployerIdOps(`rho:rchain:deployerId:ops`), deployerPubKeyBytesCh in {
+            deployerIdOps!("pubKeyBytes", *deployerID, *deployerPubKeyBytesCh) |
+            for (@pubKeyBytes <- deployerPubKeyBytesCh){
+              ops!("buildUri", pubKeyBytes, *uriCh) |
+              for (@uri <- uriCh) {
+                for (@map <<- _registryStore) {
+                  new responseCh in {
+                    TreeHashMap!("get", map, uri, *responseCh) |
+                    for (@response <- responseCh) {
+                      match response {
+                        Nil => {
+                          new ack in {
+                            TreeHashMap!("set", map, uri, (nonce, data), *ack) |
+                            for (_ <- ack) {
+                              ret!(uri)
+                            }
+                          }
+                        }
+                        (oldNonce, _) => {
+                          if (nonce > oldNonce) {
+                            new ack in {
+                              TreeHashMap!("set", map, uri, (nonce, data), *ack) |
+                              for (_ <- ack) {
+                                ret!(uri)
+                              }
+                            }
+                          } else {
+                            ret!(Nil)
+                          }
                         }
                       }
                     }
-                    (oldNonce, _) => {
-                      if (nonce > oldNonce) {
-                        new ack in {
-                          TreeHashMap!("set", map, uri, (nonce, data), *ack) |
-                          for (_ <- ack) {
-                            ret!(uri)
+                  }
+                }
+              }
+            }
+          }
+        }
+      } |
+    {
+      // ------
+      // Update
+      // deployerID: (`rho:rchain:deployerId`)
+      // uri:   the uri you want to update.  `rho:rchain:pos` etc.
+      // con:   the update consume which take the old state
+      // meth:  the target new contract which would replace the old one
+      // ret:   channel which returns the update result
+      contract systemContractManager(@"update", @securityData, @uri, @con, ret) = {
```

Comment:
Please use more descriptive names instead of _con_ and _meth_ below (it's not part of contract params).